File: src/main/java/com/pyamsoft/padlock/app/list/LockInfoItem.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.app.list;

import android.databinding.DataBindingUtil;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.RadioButton;
import com.mikepenz.fastadapter.items.GenericAbstractItem;
import com.mikepenz.fastadapter.utils.ViewHolderFactory;
import com.pyamsoft.padlock.R;
import com.pyamsoft.padlock.databinding.AdapterItemLockinfoBinding;
import com.pyamsoft.padlock.model.ActivityEntry;
import com.pyamsoft.padlock.model.LockState;
import java.lang.ref.WeakReference;
import java.util.List;
import timber.log.Timber;

class LockInfoItem extends GenericAbstractItem<ActivityEntry, LockInfoItem, LockInfoItem.ViewHolder>
    implements FilterableItem<LockInfoItem, LockInfoItem.ViewHolder> {

  @NonNull private static final ViewHolderFactory<? extends ViewHolder> FACTORY = new ItemFactory();
  @NonNull private final String packageName;

  LockInfoItem(@NonNull String packageName, @NonNull ActivityEntry entry) {
    super(entry);
    this.packageName = packageName;
  }

  @NonNull @CheckResult String getName() {
    return getModel().name();
  }

  @CheckResult @NonNull LockInfoItem copyWithNewLockState(@NonNull LockState newState) {
    return new LockInfoItem(packageName,
        ActivityEntry.builder(getModel()).lockState(newState).build());
  }

  @Override public int getType() {
    return R.id.adapter_lock_info;
  }

  @Override public int getLayoutRes() {
    return R.layout.adapter_item_lockinfo;
  }

  @Override public void bindView(ViewHolder holder, List<Object> payloads) {
    super.bindView(holder, payloads);
    holder.bind(packageName, getModel());
  }

  @Override public void unbindView(ViewHolder holder) {
    super.unbindView(holder);
    holder.unbind();
  }

  @Override public boolean filterAgainst(@NonNull String query) {
    final String name = getModel().name().toLowerCase().trim();
    Timber.d("Filter predicate: '%s' against %s", query, name);
    return !name.contains(query);
  }

  @Override public ViewHolderFactory<? extends ViewHolder> getFactory() {
    return FACTORY;
  }

  interface OnLockRadioCheckedChanged {

    void call(int position, @NonNull String name, @NonNull LockState oldState,
        @NonNull LockState newState);
  }

  @SuppressWarnings("WeakerAccess") protected static class ItemFactory
      implements ViewHolderFactory<ViewHolder> {
    @Override public ViewHolder create(View v) {
      return new ViewHolder(v);
    }
  }

  protected static final class ViewHolder extends RecyclerView.ViewHolder {

    @NonNull private final AdapterItemLockinfoBinding binding;
    @NonNull private WeakReference<ActivityEntry> weakEntry;

    ViewHolder(View itemView) {
      super(itemView);
      binding = DataBindingUtil.bind(itemView);
      weakEntry = new WeakReference<>(null);
    }

    @CheckResult @NonNull AdapterItemLockinfoBinding getBinding() {
      return binding;
    }

    void bind(@NonNull String packageName, @NonNull ActivityEntry entry) {
      // Remove any old binds
      final RadioButton lockedButton;
      switch (entry.lockState()) {
        case DEFAULT:
          lockedButton = binding.lockInfoRadioDefault;
          break;
        case WHITELISTED:
          lockedButton = binding.lockInfoRadioWhite;
          break;
        case LOCKED:
          lockedButton = binding.lockInfoRadioBlack;
          break;
        default:
          throw new IllegalStateException("Illegal enum state");
      }
      binding.lockInfoRadioBlack.setOnCheckedChangeListener(null);
      binding.lockInfoRadioWhite.setOnCheckedChangeListener(null);
      binding.lockInfoRadioDefault.setOnCheckedChangeListener(null);
      lockedButton.setChecked(true);

      final String entryName = entry.name();
      String activityName = entryName;
      if (entryName.startsWith(packageName)) {
        final String strippedPackageName = entryName.replace(packageName, "");
        if (strippedPackageName.charAt(0) == '.') {
          activityName = strippedPackageName;
        }
      }
      binding.lockInfoActivity.setText(activityName);

      weakEntry.clear();
      weakEntry = new WeakReference<>(entry);
    }

    void bind(@NonNull OnLockRadioCheckedChanged onCheckedChanged) {
      binding.lockInfoTristateRadiogroup.setOnCheckedChangeListener((radioGroup, i) -> {
        final int id = radioGroup.getCheckedRadioButtonId();
        Timber.d("Checked radio id: %d", id);
        if (id == 0) {
          Timber.e("No radiobutton is checked, set to default");
          getBinding().lockInfoRadioDefault.setChecked(true);
        }
      });
      binding.lockInfoRadioDefault.setOnCheckedChangeListener(
          new OnRadioCheckChangedListener(getAdapterPosition(), weakEntry, onCheckedChanged,
              LockState.DEFAULT));
      binding.lockInfoRadioWhite.setOnCheckedChangeListener(
          new OnRadioCheckChangedListener(getAdapterPosition(), weakEntry, onCheckedChanged,
              LockState.WHITELISTED));
      binding.lockInfoRadioBlack.setOnCheckedChangeListener(
          new OnRadioCheckChangedListener(getAdapterPosition(), weakEntry, onCheckedChanged,
              LockState.LOCKED));
    }

    void unbind() {
      binding.lockInfoActivity.setText(null);
      binding.lockInfoRadioBlack.setOnCheckedChangeListener(null);
      binding.lockInfoRadioWhite.setOnCheckedChangeListener(null);
      binding.lockInfoRadioDefault.setOnCheckedChangeListener(null);
      binding.lockInfoTristateRadiogroup.setOnCheckedChangeListener(null);
      weakEntry.clear();
    }

    static class OnRadioCheckChangedListener implements CompoundButton.OnCheckedChangeListener {

      private final int position;
      @NonNull private final WeakReference<ActivityEntry> weakEntry;
      @NonNull private final OnLockRadioCheckedChanged onCheckedChanged;
      @NonNull private final LockState changeToState;

      OnRadioCheckChangedListener(int position, @NonNull WeakReference<ActivityEntry> weakEntry,
          @NonNull OnLockRadioCheckedChanged onCheckedChanged, @NonNull LockState changeToState) {
        this.position = position;
        this.weakEntry = weakEntry;
        this.onCheckedChanged = onCheckedChanged;
        this.changeToState = changeToState;
      }

      @Override public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
        if (b) {
          final ActivityEntry entry = weakEntry.get();
          if (entry != null) {
            onCheckedChanged.call(position, entry.name(), entry.lockState(), changeToState);
          } else {
            Timber.e("Entry is NULL");
          }
        }
      }
    }
  }
}

File: padlock/src/main/java/com/pyamsoft/padlock/list/info/LockInfoItem.kt
/*
 * Copyright 2019 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.pyamsoft.padlock.list.info

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.pyamsoft.padlock.Injector
import com.pyamsoft.padlock.PadLockComponent
import com.pyamsoft.padlock.R
import com.pyamsoft.padlock.list.info.LockInfoItem.ViewHolder
import com.pyamsoft.padlock.model.LockState
import com.pyamsoft.padlock.model.list.ActivityEntry
import com.pyamsoft.padlock.model.list.ActivityEntry.Item
import com.pyamsoft.pydroid.core.bus.Publisher
import timber.log.Timber
import javax.inject.Inject

class LockInfoItem internal constructor(
  entry: ActivityEntry.Item,
  private val system: Boolean
) : LockInfoBaseItem<Item, LockInfoItem, ViewHolder>(entry) {

  override fun getType(): Int = R.id.adapter_lock_info

  override fun getLayoutRes(): Int = R.layout.adapter_item_lockinfo

  override fun bindView(
    holder: ViewHolder,
    payloads: List<Any>
  ) {
    super.bindView(holder, payloads)
    holder.bind(model, system)
  }

  override fun unbindView(holder: ViewHolder) {
    super.unbindView(holder)
    holder.unbind()
  }

  override fun filterAgainst(query: String): Boolean {
    val name = model.name.toLowerCase()
        .trim { it <= ' ' }
    Timber.d("Filter predicate: '%s' against %s", query, name)
    return name.contains(query)
  }

  override fun getViewHolder(view: View): ViewHolder =
    ViewHolder(view)

  class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {

    @field:Inject internal lateinit var publisher: Publisher<LockInfoEvent>
    @field:Inject internal lateinit var view: LockInfoItemView

    init {
      Injector.obtain<PadLockComponent>(itemView.context.applicationContext)
          .plusLockInfoItemComponent()
          .itemView(itemView)
          .build()
          .inject(this)
    }

    private fun processModifyDatabaseEntry(
      model: ActivityEntry.Item,
      system: Boolean,
      newLockState: LockState
    ) {
      publisher.publish(LockInfoEvent.from(model, newLockState, null, system))
    }

    fun bind(
      model: ActivityEntry.Item,
      system: Boolean
    ) {
      view.bind(model, system)

      view.onSwitchChanged {
        processModifyDatabaseEntry(model, system, it)
      }
    }

    fun unbind() {
      view.unbind()
    }
  }
}
